In [69]:
import pandas as pd
import numpy as np

import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

In [70]:
df = pd.read_csv('../../data/WA_Fn-UseC_-Telco-Customer-Churn.csv')

In [71]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [72]:
df.head().T

0             1               2  \
customerID              7590-VHVEG    5575-GNVDE      3668-QPYBK   
gender                      Female          Male            Male   
SeniorCitizen                    0             0               0   
Partner                        Yes            No              No   
Dependents                      No            No              No   
tenure                           1            34               2   
PhoneService                    No           Yes             Yes   
MultipleLines     No phone service            No              No   
InternetService                DSL           DSL             DSL   
OnlineSecurity                  No           Yes             Yes   
OnlineBackup                   Yes            No             Yes   
DeviceProtection                No           Yes              No   
TechSupport                     No            No              No   
StreamingTV                     No            No              No   
StreamingMovies                 No            No              No   
Contract            Month-to-month      One year  Month-to-month   
PaperlessBilling               Yes            No             Yes   
PaymentMethod     Electronic check  Mailed check    Mailed check   
MonthlyCharges               29.85         56.95           53.85   
TotalCharges                 29.85        1889.5          108.15   
Churn                           No            No             Yes   

                                          3                 4  
customerID                       7795-CFOCW        9237-HQITU  
gender                                 Male            Female  
SeniorCitizen                             0                 0  
Partner                                  No                No  
Dependents                               No                No  
tenure                                   45                 2  
PhoneService                             No               Yes  
MultipleLines              No phone service                No  
InternetService                         DSL       Fiber optic  
OnlineSecurity                          Yes                No  
OnlineBackup                             No                No  
DeviceProtection                        Yes                No  
TechSupport                             Yes                No  
StreamingTV                              No                No  
StreamingMovies                          No                No  
Contract                           One year    Month-to-month  
PaperlessBilling                         No               Yes  
PaymentMethod     Bank transfer (automatic)  Electronic check  
MonthlyCharges                         42.3              70.7  
TotalCharges                        1840.75            151.65  
Churn                                    No               Yes

In [73]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df['TotalCharges'] = df['TotalCharges'].fillna(0)

In [74]:
df['SeniorCitizen'] = df['SeniorCitizen'].map({0: 'no', 1: 'yes'})


In [75]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
string_columns = list(df.dtypes[df.dtypes == 'object'].index)

for col in string_columns:
    df[col] = df[col].str.lower().str.replace(' ', '_')

In [76]:
df.churn = (df.churn == 'yes').astype(int)

In [77]:
from sklearn.model_selection import train_test_split

In [78]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=1)

In [79]:
df_train, df_val = train_test_split(df_train_full, test_size=0.33, random_state=11)

In [80]:
y_train = df_train.churn.values
y_val = df_val.churn.values

In [81]:
del df_train['churn']
del df_val['churn']

In [82]:
categorical = ['gender', 'seniorcitizen', 'partner', 'dependents',
               'phoneservice', 'multiplelines', 'internetservice',
               'onlinesecurity', 'onlinebackup', 'deviceprotection',
               'techsupport', 'streamingtv', 'streamingmovies',
               'contract', 'paperlessbilling', 'paymentmethod']
numerical = ['tenure', 'monthlycharges', 'totalcharges']

In [83]:
from sklearn.feature_extraction import DictVectorizer

dv  = DictVectorizer(sparse=False)

X_train_vec = df_train[categorical + numerical].to_dict(orient = 'records')

dv.fit(X_train_vec)

X_train = dv.transform(X_train_vec)

In [84]:
dv = DictVectorizer(sparse=False)

X_val_vec = df_val[categorical + numerical].to_dict(orient = 'records')

dv.fit(X_val_vec)

X_val = dv.transform(X_val_vec)

In [85]:
X_val.shape

(1860, 46)

In [86]:
dv.get_feature_names_out()

array(['contract=month-to-month', 'contract=one_year',
       'contract=two_year', 'dependents=no', 'dependents=yes',
       'deviceprotection=no', 'deviceprotection=no_internet_service',
       'deviceprotection=yes', 'gender=female', 'gender=male',
       'internetservice=dsl', 'internetservice=fiber_optic',
       'internetservice=no', 'monthlycharges', 'multiplelines=no',
       'multiplelines=no_phone_service', 'multiplelines=yes',
       'onlinebackup=no', 'onlinebackup=no_internet_service',
       'onlinebackup=yes', 'onlinesecurity=no',
       'onlinesecurity=no_internet_service', 'onlinesecurity=yes',
       'paperlessbilling=no', 'paperlessbilling=yes', 'partner=no',
       'partner=yes', 'paymentmethod=bank_transfer_(automatic)',
       'paymentmethod=credit_card_(automatic)',
       'paymentmethod=electronic_check', 'paymentmethod=mailed_check',
       'phoneservice=no', 'phoneservice=yes', 'seniorcitizen=no',
       'seniorcitizen=yes', 'streamingmovies=no',
       'stream

In [87]:
# Now Apply The SMOTE
from imblearn.over_sampling import SMOTE
oversample = SMOTE()

X_train_smote, y_train_smote = oversample.fit_resample(X_train, y_train)
X_val_smote, y_val_smote = oversample.fit_resample(X_val, y_val)

In [101]:
import pickle
import mlflow
from sklearn.model_selection import GridSearchCV
from xgboost import XGBClassifier

mlflow.set_tracking_uri("sqlite:///mlflow.db")
mlflow.set_experiment("Telecom Churn Prediction 4")

<Experiment: artifact_location=('file:///c:/Users/micheal.getachew/OneDrive - Safaricom '
 'Ethiopia/Desktop/Projects/Churn_Prediction_Model/notebooks/tracked/mlruns/4'), creation_time=1688939271668, experiment_id='4', last_update_time=1688939271668, lifecycle_stage='active', name='Telecom Churn Prediction 4', tags={}>

In [102]:
from sklearn.model_selection import cross_val_score
from hyperopt import fmin, tpe, hp, STATUS_OK, Trials
from hyperopt.pyll import scope

In [103]:
import xgboost as xgb
train = xgb.DMatrix(X_train_smote, label=y_train_smote)
valid = xgb.DMatrix(X_val_smote, label=y_val_smote)

In [ ]:
# This was my first trial : -)
def objective(params):
    with mlflow.start_run():
        mlflow.set_tag("model", "xgboost")
        mlflow.log_params(params)
        clf = XGBClassifier(**params)
        score = cross_val_score(clf, X_train, y_train, cv=5).mean()
        mlflow.log_metric("score", score)
    return {'loss': score, 'status': STATUS_OK}

space = {
    'max_depth': hp.choice('max_depth', range(1, 20)),
    'learning_rate': hp.loguniform('learning_rate', -5, 0),
    'n_estimators': hp.choice('n_estimators', range(1, 1000)),
    'gamma': hp.loguniform('gamma', -5, 0),
    'min_child_weight': hp.choice('min_child_weight', range(1, 10)),
    'subsample': hp.uniform('subsample', 0.1, 1),
    'colsample_bytree': hp.uniform('colsample_bytree', 0.1, 1),
    'reg_alpha': hp.uniform('reg_alpha', 0.01, 100),
    'reg_lambda': hp.uniform('reg_lambda', 0.01, 100)
}

best = fmin(fn=objective,
            space=space,
            algo=tpe.suggest,
            max_evals=100)

print(best)

# def objective(params):
#     with mlflow.start_run():
#         mlflow.set_tag("model", "xgboost")
#         mlflow.log_params(params)
#         booster = xgb.train(
#             params=params,
#             dtrain=train,
#             num_boost_round=1000,
#             evals=[(valid, 'validation')],
#             early_stopping_rounds=50
#         )
#         y_pred = booster.predict(valid)
#         rmse = mean_squared_error(y_val, y_pred, squared=False)
#         mlflow.log_metric("rmse", rmse)

#     return {'loss': rmse, 'status': STATUS_OK}

In [104]:
import xgboost as xgb
from hyperopt import fmin, tpe, hp

# Define search space
space = {
    'max_depth': hp.choice('max_depth', range(1, 20)),
    'learning_rate': hp.quniform('learning_rate', 0.01, 0.5, 0.01),
    'n_estimators': hp.choice('n_estimators', range(1, 1000)),
    'gamma': hp.quniform('gamma', 0, 1, 0.01),
    'min_child_weight': hp.quniform('min_child_weight', 1, 10, 1),
    'subsample': hp.quniform('subsample', 0.1, 1, 0.01),
    'colsample_bytree': hp.quniform('colsample_bytree', 0.1, 1.0, 0.01)
}

# Define objective function
def objective(params):
    with mlflow.start_run():
        mlflow.set_tag("Model", "Xgboost")
        mlflow.log_params(params)
        clf = xgb.XGBClassifier(**params)
        clf.fit(X_train_smote, y_train_smote)
        score = clf.score(X_val_smote, y_val_smote)
        mlflow.log_metric("score", score)
    return {'loss': -score, 'status': STATUS_OK}

# Run hyperparameter search
best = fmin(fn=objective,
            space=space,
            algo=tpe.suggest,
            max_evals=100,
            trials=Trials())


100%|██████████| 100/100 [02:43<00:00,  1.63s/trial, best loss: -0.8584425036390102]


In [110]:
# As it can be seen from the previous cell we can confirm that the accuracy of the best model is around 85.8% thus we can save it for next steps

# Train final model with best hyperparameters

mlflow.xgboost.autolog()
clf = xgb.XGBClassifier(**best)
clf.fit(X_train, y_train)

with open("./preprocessor.b", "wb") as f_out:
    pickle.dump(dv, f_out)

mlflow.log_artifact("./preprocessor.b", artifact_path="preprocessor")

2023/07/09 15:46:54 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of xgboost. If you encounter errors during autologging, try upgrading / downgrading xgboost to a supported version, or try upgrading MLflow.
2023/07/09 15:46:55 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\micheal.getachew\AppData\Local\anaconda3\envs\churn\lib\site-packages\xgboost\sklearn.py:761: UserWarning: n_estimators is not saved in Scikit-Learn meta."
2023/07/09 15:46:55 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\micheal.getachew\AppData\Local\anaconda3\envs\churn\lib\site-packages\xgboost\sklearn.py:761: UserWarning: max_depth is not saved in Scikit-Learn meta."
2023/07/09 15:46:58 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\micheal.getachew\AppData\Local\anaconda3\envs\churn\lib\site-packages\_distutils_hack\__init__.py:18: UserWarning: Di

In [111]:
# Check how the model was saved
import mlflow
logged_model = 'runs:/7a5716aeb16a4182ba4cb0e5b6943be3/model'

# Load model as a PyFuncModel.
loaded_model = mlflow.pyfunc.load_model(logged_model)

2023/07/09 15:50:39 WARNING mlflow.pyfunc: Detected one or more mismatches between the model's dependencies and the current Python environment:
 - mlflow (current: 2.4.1, required: mlflow==2.4)
To fix the mismatches, call `mlflow.pyfunc.get_model_dependencies(model_uri)` to fetch the model's environment and install dependencies using the resulting environment file.


In [112]:
loaded_model

mlflow.pyfunc.loaded_model:
  artifact_path: model
  flavor: mlflow.xgboost
  run_id: 7a5716aeb16a4182ba4cb0e5b6943be3

In [113]:
# Or we can load the model as an xgboost object like ...
xgboost_model = mlflow.xgboost.load_model(logged_model)

In [114]:
xgboost_model

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.25, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=0.43, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.07, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=2.0, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, ...)

In [115]:
X_test_vectorized = df_test[categorical + numerical].to_dict(orient = 'records')

X_test = dv.transform(X_test_vectorized)

In [127]:
X_test.shape

(1409, 46)

In [116]:
y_pred = xgboost_model.predict(X_test)

In [118]:
y_pred[:100]

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1])

In [119]:
y_test = df_test.churn.values

In [120]:
y_test[:100]

array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [125]:
count = 0
for i in range(len(y_test)):
    if y_test[i] == y_pred[i]:
        count += 1
    else:
        continue

print(f"The Match is: {count/len(y_test)}")

The Match is: 0.8119233498935415


#### <a>As it can be seen from the above output our model have predicted 81.2% of the unseen data correctly !!!</a>

In [126]:
# Save the model in pickle for latter
import pickle

with open('model.pkl', 'wb') as f:
    pickle.dump(xgboost_model, f)


In [128]:
# Save the test dataset as a csv to test how our api is doing
df_test = df_test[categorical + numerical]
df_test.to_csv('test.csv', index=False)

In [144]:
import requests
import json
url = 'http://127.0.0.1:9696/predict'

#json_data = X_test.to_json(orient='records')
json_data = json.dumps(np.ndarray.tolist(X_test))
response = requests.post(url, json=json_data)
print(response.json())
#print(json_data)

{'predictions': [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0,

#### <a>All is fine now</a>